EmojicodeReal-TimeEngine/EmojicodeDictionary.c

```diff
@@ -10,121 +10,342 @@
 #include "EmojicodeDictionary.h"
 #include "EmojicodeString.h"
 
-#define FNV_PRIME_32 16777619
-#define FNV_OFFSET_32 2166136261U
-uint32_t fnv32(const char *k, size_t length){
-    uint32_t hash = FNV_OFFSET_32;
+#define FNV_PRIME_64 1099511628211
+#define FNV_OFFSET_64 14695981039346656037U
+EmojicodeDictionaryHash fnv64(const char *k, size_t length) {
+    EmojicodeDictionaryHash hash = FNV_OFFSET_64;
     for(size_t i = 0; i < length; i++){
         hash = hash ^ k[i];
-        hash = hash * FNV_PRIME_32;
+        hash = hash * FNV_PRIME_64;
     }
     return hash;
 }
 
-#define slots(dict) ((EmojicodeDictionaryKVP *)(dict->slots->value))
+EmojicodeDictionaryHash dictionaryHash(EmojicodeDictionary *dict, Something key) {
+    // TODO: Implement hash Callable
+    #define hashString(keyString) fnv64((char*)characters(keyString), ((keyString)->length) * sizeof(EmojicodeChar))
+    return hashString((String *) key.object->value);
+}
+
+bool dictionaryKeyEqual(EmojicodeDictionary *dict, Something key1, Something key2) {
+    return stringEqual((String *) key1.object->value, (String *) key2.object->value);
+}
 
+bool dictionaryKeyHashEqual(EmojicodeDictionary *dict, EmojicodeDictionaryHash hash1, EmojicodeDictionaryHash hash2, Something key1, Something key2) {
+    return (hash1 == hash2) && dictionaryKeyEqual(dict, key1, key2);
+}
+
+// MARK: Internal dictionary
+EmojicodeDictionaryNode* dictionaryGetNode(EmojicodeDictionary *dict, EmojicodeDictionaryHash hash, Something key) {
+    EmojicodeDictionaryNode *e;
+    Object** bucko;
+    size_t n = 0;
+    if (dict->buckets != NULL) {
+        bucko = (Object**) dict->buckets->value;
+        if ((n = dict->bucketsCounter) > 0) {
+            Object *firsto = bucko[hash & (n - 1)];
+            if (firsto != NULL) {
+                e = firsto->value;
+                if (dictionaryKeyHashEqual(dict, hash, e->hash, key, e->key)) {
+                    return e;
+                }
+                Object *eo;
+                while ((eo = e->next)) {
+                    e = eo->value;
+                    if (dictionaryKeyHashEqual(dict, hash, e->hash, key, e->key)) {
+                        return e;
+                    }
+                }
+            }
+        }
+    }
+    return NULL;
+}
 
-void dictionaryInit(Thread *thread){
-#define initialCapactiy 7
-    Object *slots = newArray(sizeof(EmojicodeDictionaryKVP) * (initialCapactiy));
+Object* dictionaryNewNode(Object **dicto, EmojicodeDictionaryHash hash, Something key, Something value, Object *next, Thread *thread){
+    stackPush(*dicto, 0, 0, thread);
```

Comment:
Maybe you want the callee to take care of this as the callees anyhow need to "keep a strong reference" to `dicto`.